Create users at install time for image and edge installers

achilleas-k opened the pull request:
Rewrite of https://github.com/osbuild/osbuild-composer/pull/1700 for RHEL 8.6 and RHEL 9.0 (and CentOS 8 and CentOS 9).

Instead of "baking in" users to the payload, create them at install time.  For the edge-installer, this has the benefit that we no longer need to work around the ssh key creation on first boot.

For the image-installer, this is functionally equivalent.

The edge-installer previously did not allow any customizations at build time.  Now users and groups are allowed.

Added test cases to generate manifests for the image- and edge-installer with users.

teg commented:
This is really great! Thanks for picking it up :)

ondrejbudai reviewed (dismissed):
Needs an exception I think.

achilleas-k commented:
> @achilleas-k Is this PR ready for testing? May I add edge-installer test in this PR? Thanks!

Yes, please do.  I was trying to figure out where the failure is coming from, but maybe we should test the feature first and see how that goes.

achilleas-k commented:
Pushed a rebase on `main`

henrywang commented:
@achilleas-k I added `edge-installer` test, it passed. But I found all `installer` tests failed. After installation, vm entered into emergency shell. Could you please take a look if you have time? Thanks.
```
[    1.774983] localhost dracut-mount[490]: Warning: Can't mount root filesystem
[    1.793486] localhost systemd[1]: Starting Dracut Emergency Shell...
```

```
+ cat /proc/self/mountinfo
1 1 0:2 / / rw shared:1 - rootfs rootfs rw,size=1453724k,nr_inodes=363431,inode64
22 1 0:21 / /proc rw,nosuid,nodev,noexec,relatime shared:2 - proc proc rw
23 1 0:22 / /sys rw,nosuid,nodev,noexec,relatime shared:3 - sysfs sysfs rw
24 1 0:5 / /dev rw,nosuid shared:8 - devtmpfs devtmpfs rw,size=1453744k,nr_inodes=363436,mode=755,inode64
25 23 0:6 / /sys/kernel/security rw,nosuid,nodev,noexec,relatime shared:4 - securityfs securityfs rw
26 24 0:23 / /dev/shm rw,nosuid,nodev shared:9 - tmpfs tmpfs rw,inode64
27 24 0:24 / /dev/pts rw,nosuid,noexec,relatime shared:10 - devpts devpts rw,gid=5,mode=620,ptmxmode=000
28 1 0:25 / /run rw,nosuid,nodev shared:11 - tmpfs tmpfs rw,size=595368k,nr_inodes=819200,mode=755,inode64
29 23 0:26 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:5 - cgroup2 cgroup2 rw,nsdelegate,memory_recursiveprot
30 23 0:27 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime shared:6 - pstore pstore rw
31 23 0:28 / /sys/fs/bpf rw,nosuid,nodev,noexec,relatime shared:7 - bpf none rw,mode=700
32 1 0:29 / /var/lib/nfs/rpc_pipefs rw,relatime shared:12 - rpc_pipefs rpc_pipefs rw
+ cat /proc/mounts
rootfs / rootfs rw,size=1453724k,nr_inodes=363431,inode64 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
devtmpfs /dev devtmpfs rw,nosuid,size=1453744k,nr_inodes=363436,mode=755,inode64 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev,inode64 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,nodev,size=595368k,nr_inodes=819200,mode=755,inode64 0 0
cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
none /sys/fs/bpf bpf rw,nosuid,nodev,noexec,relatime,mode=700 0 0
rpc_pipefs /var/lib/nfs/rpc_pipefs rpc_pipefs rw,relatime 0 0
```

jrusz commented:
I just rebased on top of `main`

achilleas-k commented:
Exceptions granted for [8.6](https://bugzilla.redhat.com/show_bug.cgi?id=1951936) and [9.0](https://bugzilla.redhat.com/show_bug.cgi?id=2060575).

Rebased on `main`.

achilleas-k commented:
All the edge-installer tests are passing here.  The only failure is the 8.6 installer test.

achilleas-k commented:
The [latest commit](https://github.com/osbuild/osbuild-composer/pull/2375/commits/[3a17a4d](https://github.com/osbuild/osbuild-composer/pull/2375/commits/3a17a4d4946872d44a5e95de1b825459bfdb7879)4946872d44a5e95de1b825459bfdb7879) works around the installer issue by adding `restorecon` to the end of the kickstart file.  I think the issue we're facing might be related to the modifications we make to the kickstart file.  When I run the installer interactively locally, the issue doesn't occur.  I'm going to investigate further, but I think we can keep the workaround to avoid missing the exception deadline.

achilleas-k commented:
> The [latest commit](https://github.com/osbuild/osbuild-composer/pull/2375/commits/3a17a4d4946872d44a5e95de1b825459bfdb7879) works around the installer issue by adding `restorecon` to the end of the kickstart file. I think the issue we're facing might be related to the modifications we make to the kickstart file. When I run the installer interactively locally, the issue doesn't occur. I'm going to investigate further, but I think we can keep the workaround to avoid missing the exception deadline.

Reverted this and added a proper, permanent workaround.  See commit for details.

achilleas-k commented:
http://download.eng.bos.redhat.com/rhel-8/nightly/RHEL-8/latest-finished-RHEL-8.6/COMPOSE_ID doesn't exist for https://github.com/achilleas-k/osbuild-composer/blob/08ecc6020f3bfa6e351f2e74c689c84583f3bc0c/tools/define-compose-url.sh#L11

thozza reviewed (commented):
Great stuff!

I added a few comments. None of them are critical, but I would like to wait for their resolution before approving.

I love the de-duplication of various functions and their move to the `osbuild2` package. I would suggest to de-duplicate these functions also in older distro definitions. Although some of them would slightly change manifests, those changes have no effect on the functionality and backward compatibility of these images. This should be completely OK as an exception to our policy. The policy should IMO not prevent us from doing "the right thing" and cleaning up old code. If we do not want to maintain these definitions for the next 10 years as they are currently implemented, but want to eventually replace them with their equivalents using new code, then we should be cleaning them up and de-duplicating the code in them along the way.

In addition, I do not think that we should be blocking the merge of this PR due to waiting on RHEL exception. The reason is that we can selectively backport specific commits to RHEL or cherry-pick them on some other branch in upstream for that release. Blocking development changes on `main` for this reason IMO does not make sense.

achilleas-k commented:
With respect to deduplicating calls and changing manifests in other distros, since this is specifically about an exception we want to get into 8.6 and 9.0, I wanted to minimise any side-effects to other image types and distributions.  The priority here is getting the user-creation to be a bit more sane to fix the relevant bug and to minimise potential issues during the lifetime of the next version.

achilleas-k commented on internal/distro/rhel90/pipelines.go:
```diff
@@ -407,28 +407,16 @@ func osPipeline(t *imageType,
-		if err != nil {
-			return nil, err
-		}
+	if userOptions, err := osbuild.NewUsersStageOptions(c.GetUsers(), false); err != nil {
```
I think that would confuse the purpose of the bool value.  The point is to get the user options with or without the key included.  For ostree-based images we actually need both without and with keys, the former for the kickstart stage and the latter for the firstboot stage.  Using the `t.rpmOstree` flag for one call and then ignoring it in the next would make the meaning of the option confusing, I believe.

thozza replied:
Makes sense. I missed that the `userOptions` is actually passed to the firstboot stage.

achilleas-k commented on internal/osbuild2/groups_stage.go:
```diff
@@ -1,13 +1,17 @@
 
 type GroupsStageOptionsGroup struct {
-	Name string `json:"name"`
+	// NOTE: the name field does not exist in the osbuild schema and should be
+	// removed
+	Name string `json:"name,omitempty"`
```
See my PR comment about minimising side-effects.

thozza commented:
> With respect to deduplicating calls and changing manifests in other distros, since this is specifically about an exception we want to get into 8.6 and 9.0, I wanted to minimise any side-effects to other image types and distributions. The priority here is getting the user-creation to be a bit more sane to fix the relevant bug and to minimise potential issues during the lifetime of the next version.

I understand why you didn't do it.

However I'm not sure about the implied consequences of your comment, since the deduplication could have been done in a separate commit, which would not be backported to 8.6 and 9.0.

Does this mean that you still plan to deduplicate the code in older distros eventually in a separate PR? Or not at all?

achilleas-k commented on internal/osbuild2/kickstart_stage.go:
```diff
@@ -33,3 +37,48 @@ func NewKickstartStage(options *KickstartStageOptions) *Stage {
+func NewKickstartStageOptions(
+	path string,
+	imageURL string,
+	userCustomizations []blueprint.UserCustomization,
+	groupCustomizations []blueprint.GroupCustomization,
```
I suppose we could, but maybe if we end up needing more? I think I like that it's explicit what's needed, though I see your point of maybe having future customizations that might become relevant to the kickstart stage.

thozza replied:
👍  I agree with waiting for a situation when we would need to add additional customization value.

thozza reviewed (approved):
It turns out that there is a time pressure to get a working version of these changes in, so that they can be backported to 8.6 and 9.0.

As I mentioned, none of my comments are critical. So I'm approving this PR with the expectation that the raised points will be considered for a follow-up PR.

achilleas-k commented on internal/distro/rhel86/distro.go:
```diff
@@ -587,8 +588,11 @@ func (t *imageType) checkOptions(customizations *blueprint.Customizations, optio
 			}
-		} else if customizations != nil {
-			return fmt.Errorf("boot ISO image type %q does not support blueprint customizations", t.name)
+		} else if t.name == "edge-installer" {
```
Image installer has no blueprint restrictions so it's never checked.  Edge installer used to not allow any blueprint customisations so now it's checked to only allow users and groups.

thozza replied:
I think I just realized why and that you are correct. Probably just my over-confusion from endless regeneration of image test cases 😕
